From e1fa6b18c0c879f4e85cb04753ff409077664e5b Mon Sep 17 00:00:00 2001
From: Ameermon V A <ameermonx.va@intel.com>
Date: Thu, 12 May 2016 17:07:53 +0530
Subject: [PATCH 1/1] EM/dc_ti_pwrsrc: WA fix to enable the VBUS upon wakeup
 from S3

Issue:
When the USB peripheral device is connected to the platform
during S3, platform got the IRQ and pwrsrc driver got USB-Host
extcon notification. Soon after, pwrsrc driver is trying to
enable the VBUS through bq charger chip register. But the i2c
bus in which the bq chip is placed has not come out of suspend
at this time. This results in i2c read-write error.

Solution:
Added a delay of 10ms to start the worker thread which enables
the VBUS. By this time i2c bus has completed resume function
and so bq chip access and the VBUS enable is happening
successfully.

Change-Id: I0f97997964dbed82196a86149c17f1f146065714
Signed-off-by: Ameermon V A <ameermonx.va@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-30850
Reviewed-on: https://android.intel.com:443/505552
---
 drivers/extcon/extcon-dc-ti-pwrsrc.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/extcon/extcon-dc-ti-pwrsrc.c b/drivers/extcon/extcon-dc-ti-pwrsrc.c
index 706d393..7b2199e 100644
--- a/drivers/extcon/extcon-dc-ti-pwrsrc.c
+++ b/drivers/extcon/extcon-dc-ti-pwrsrc.c
@@ -235,7 +235,14 @@ static int dc_ti_pwrsrc_handle_extcon_event(struct notifier_block *nb,
 	struct dc_pwrsrc_info *info =
 	    container_of(nb, struct dc_pwrsrc_info, extcon_nb);
 
-	schedule_delayed_work(&info->dc_pwrsrc_wrk, 0);
+	/*
+	 * After getting "USB-Host" extcon event notification, it is
+	 * taking few milli seconds to resume the i2c bus from suspend
+	 * where bq charger chip is placed. So giving a 10 ms
+	 * delay before scheduling dc_pwrsrc_wrk which infact calls
+	 * the vbus enable function.
+	 */
+	schedule_delayed_work(&info->dc_pwrsrc_wrk, msecs_to_jiffies(10));
 
 	return NOTIFY_OK;
 }
-- 
1.9.1

